### DISCUSSION TITLE
CI/CD / GitOps?

### DISCUSSION THREAD

**Discussion by PehrGit** [htpcBeginner/docker-traefik#82] [Ideas]
Hi,

I'm building a new home server as replacement for my NAS. I've had my eye on this repo for a while now, looks really awesome, kudos! The main reason it appeals to me is that I can put the configuration of the different services into source control, giving me:
- If the machine dies I have a backup of the config of all my apps.
- If I ever want to reinstall the OS, I don't have to install and configure the apps one by one.
- I have version history so I can keep track of the different configs that I have tried.

So putting the docker-compose.yml in a git repo, cloning that, and running "up" will work. But it would be even *more* awesome if I could keep the machine zero-touch and make sure that I don't change things by hand which I don't document and will forget to check into git.

Also, not having to SSH into the machine to update the config when I add or change the config of an app would be super nice.

As a dev I've grown used to having my git changes be integrated and deployed continuously (CI/CD). I was also inspired by the concept of GitOps (which I recently learned about), I'm looking for a solution where I can push configuration changes to my git repo and have them be automatically applied on the server. 

Has anyone set up this kind of a CI/CD / GitOps pipeline? 

Ideally, I would not have to allow any outside-in traffic to get this working. So running SSH from a Jenkins pipeline is not what I'm looking for. It would be better if some process on the machine would watch for changes in git and apply the config. 

The main environment where GitOps tools work is in a Kubernetes cluster. But running a Kubernetes cluster on the machine seems a bit overkill. The only implementation of GitOps for Docker Compose I could find was this very basic solution: https://github.com/zeeke/docker-compose-gitops. 

The above tool will probably work but I'm still curious how others have approached this. I'm open to any kind of suggestions so if you have any thoughts, shoot!

**Comment by SimpleHomelab** [comment]
GitOps is over my head. Moving this to discussions as this is not an issue.
